Added attribute to store UID ranges

metodej opened the pull request:
- added new attributes to store UID ranges instead of UID max/min
- renamed method used to check and convert GID and now UID ranges

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java:
```diff
@@ -1085,24 +1085,24 @@ public User getUserFromMessage(PerunSessionImpl sess, String message) throws Int
-			if(maximumOfRange == null || maximumOfRange.isEmpty()) throw new WrongAttributeValueException(gidRangesAttribute, "Maximum in one of gid ranges is empty!");
+			if(minimumOfRange == null || minimumOfRange.isEmpty()) throw new WrongAttributeValueException(idRangesAttribute, "Minimum in one of gid ranges is empty!");
+			String maximumOfRange = idRanges.get(minimumOfRange);
+			if(maximumOfRange == null || maximumOfRange.isEmpty()) throw new WrongAttributeValueException(idRangesAttribute, "Maximum in one of gid ranges is empty!");
```
In the text there should be 'one of id ranges'.

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java:
```diff
@@ -1128,7 +1128,7 @@ public User getUserFromMessage(PerunSessionImpl sess, String message) throws Int
 		Integer lastMaxValue = 0;
 		for(Integer minValue: convertedRanges.keySet().stream().sorted().collect(Collectors.toList())) {
-			if(minValue <= lastMaxValue) throw new WrongAttributeValueException(gidRangesAttribute, "There is an overlap between two gid ranges.");
+			if(minValue <= lastMaxValue) throw new WrongAttributeValueException(idRangesAttribute, "There is an overlap between two gid ranges.");
```
Text should be 'between two id ranges'.

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:
```diff
@@ -0,0 +1,108 @@
+	@Override
+	public Attribute fillAttribute(PerunSessionImpl sess, Facility facility, AttributeDefinition attributeDefinition) {
+		return new Attribute(attributeDefinition);
+	}
```
I don't think you need to override this method if there is no change in it.

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:
```diff
@@ -0,0 +1,108 @@
+			return sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, uidNamespace, A_E_namespaceUIDRanges);
+		} catch(AttributeNotExistsException ex) { throw new ConsistencyErrorException(ex);
+		} catch(WrongAttributeAssignmentException ex) { throw new InternalErrorException(ex);
+		}
```
This is not wrong, but I would prefer to have a throw on separate line, please.

stavamichal commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:
```diff
@@ -0,0 +1,108 @@
+			throw new WrongReferenceAttributeValueException(attribute, namespaceUIDRangesAttribute, "You can't change attribute value of UID Ranges by changing value of virtual attribute.");
+		}
+		return false;
+	}
```
Maybe it is problem with gid ranges too, but I am not sure why is this logic here at all. You definitely can't set anything this way (there is no logic for setting in it) so maybe we should skip it at all. Or did I miss something?

metodej commented:
Thank you for review. I corrected the leftover "gid" in comments of exceptions, I've also added test for one of the modules, the other one just uses one method from ModulesUtils... I've deleted the setAttribute method, I took it from gidRanges attribute but I don't see any reason for it, sorry about that.

zlamalp commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:
```diff
@@ -0,0 +1,93 @@
+	private static final String A_FAC_uidNamespace = AttributesManager.NS_FACILITY_ATTR_DEF + ":uid-namespace";
+
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl sess, Facility facility, Attribute attribute) throws InternalErrorException, WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
```
We no longer use ``InternalErrorException`` in the API, please remove it from all the methods.

metodej replied:
Okay, removed.

metodej commented on perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:
```diff
@@ -0,0 +1,93 @@
+		attr.setNamespace(AttributesManager.NS_FACILITY_ATTR_VIRT);
+		attr.setFriendlyName("UIDRanges");
+		attr.setDisplayName("UID ranges in set namespace for the Facility");
+		attr.setType(List.class.getName());
```
It has the correct type now, thank you.
